enhancements/20240604-lvm-csi-driver.md

```diff
@@ -0,0 +1,81 @@
+
+### User Stories
+
+#### The use needs high performance and live migration is not required.
+Sometimes, the user needs a high-performance PVC, and Live Migration is not required. The user could consider using the LVM volume. The LVM volume could provide higher performance, but it can not be migrated to another node. So, if the user fits the above scenario, the user can get a better performance volume with the LVM volume.
+
+### User Experience In Detail
+
+The user creates a VM and starts with the feature with PCI pass-through. The VM is not migratable because the user uses the PCI pass-through feature. Then, the user can create the PVC with the LVM volume to get better performance. The user attaches the PVC to the VM through hotplug and uses the LVM volume in the VM.
+
+### API changes
+TBD
+
+## Design
+
+### Implementation Overview
+
+A overview of the design could be checked below:
+![lvm-csi-driver-arch](20240604-lvm-csi-driver/lvm-csi-driver-arch.png)
+
+### LVM CSI Driver
+At first, we consider two existing solutions: [metal-stack/csi-driver-lvm](https://github.com/metal-stack/csi-driver-lvm) and [topolvm/topolvm](https://github.com/topolvm/topolvm). Both solutions support the CSI topology feature, and other features like volume capability and snapshot are only supported in the topolvm. However, the topolvm is more complex than the csi-driver-lvm. We could try to add the capability and snapshot feature to the csi-driver-lvm and keep it as simple as we want.
+
+So, we chose the [metal-stack/csi-driver-lvm](https://github.com/metal-stack/csi-driver-lvm) as the base LVM CSI driver. This project is lightweight and already has the essential capability for our requirements. We want to introduce it in the first version. Then, we can add more features based on the requirements.
+
+The LVM CSI driver includes the CSI plugin and the LVM provisioner. The CSI plugin handles all CSI requests, while the LVM provisioner manages LVM operations such as creation, deletion, expansion, and snapshotting. This division of responsibilities ensures efficient and effective management of our storage infrastructure.
+
+In the initial version, we propose a significant change that will enhance user flexibility. We plan to allow users to predefine the volume group. The provisioner will then use this predefined volume group and logical volume type (e.g., mirror, stripe) to create the volume. This change will enable users to group similar types of disks together, thereby ensuring optimal performance and capability.
+
+With this change, the StorageClass will now contain two additional parameters: `vgName` and `lvType`. The `vgName` refers to the volume group name, and the `lvType` specifies the logical volume type. These parameters are essential for the LVM provisioner to create the volume according to the user's specifications.
```

Comment:
>and the `lvType` specifies the logical volume type.

Also type-specific parameters. We need to list all supported types and their parameter in the HEP.

Reply:
Do we need to specify the `nodeName` here? Or do all nodes containing a VG with the name provide volume provisioning?

Reply:
>Also type-specific parameters. We need to list all supported types and their parameter in the HEP.

Updated, thanks!

>Do we need to specify the nodeName here? Or do all nodes containing a VG with the name provide volume provisioning?

I am still thinking about that. Currently, we will create the LV with the corresponding node affinity.
If the node had the same vgName, the LV could be created on there. But that will introduce another issue, we did not support PVC migrate (no matter online or offline).

If we did not specific the nodeName, I am afraid some user will get confused.